src/python/grpcio_tests/tests_py3_only/interop/xds_interop_server.py

```diff
@@ -0,0 +1,170 @@
+from typing import DefaultDict, Dict, List, Mapping, Set, Sequence, Tuple
+import collections
+
+from concurrent import futures
+
+import grpc
+from grpc_channelz.v1 import channelz
+from grpc_channelz.v1 import channelz_pb2
+from grpc_health.v1 import health_pb2, health_pb2_grpc
+from grpc_health.v1 import health as grpc_health
+from grpc_reflection.v1alpha import reflection
+
+from src.proto.grpc.testing import test_pb2
+from src.proto.grpc.testing import test_pb2_grpc
+from src.proto.grpc.testing import messages_pb2
+from src.proto.grpc.testing import empty_pb2
+
+# NOTE: This interop server is not fully compatible with all xDS interop tests.
+#  It currently only implements enough functionality to pass the xDS security
+#  tests.
+
+_LISTEN_HOST = "[::]"
+
+_THREAD_POOL_SIZE = 256
+
+logger = logging.getLogger()
+console_handler = logging.StreamHandler()
+formatter = logging.Formatter(fmt='%(asctime)s: %(levelname)-8s %(message)s')
+console_handler.setFormatter(formatter)
+logger.addHandler(console_handler)
+
+
+class TestService(test_pb2_grpc.TestServiceServicer):
+
+    def __init__(self, server_id, hostname):
+        self._server_id = server_id
+        self._hostname = hostname
+
+    def EmptyCall(self, _: empty_pb2.Empty,
+                  context: grpc.ServicerContext) -> empty_pb2.Empty:
+        return empty_pb2.Empty()
+
+    def UnaryCall(self, request: messages_pb2.SimpleRequest,
+                  context: grpc.ServicerContext) -> messages_pb2.SimpleResponse:
+        response = messages_pb2.SimpleResponse()
+        response.server_id = self._server_id
+        response.hostname = self._hostname
+        return response
+
+
+def _configure_maintenance_server(server: grpc.Server,
+                                  maintenance_port: int) -> None:
+    channelz.add_channelz_servicer(server)
+    listen_address = f"{_LISTEN_HOST}:{maintenance_port}"
+    server.add_insecure_port(listen_address)
+    health_servicer = grpc_health.HealthServicer(
+        experimental_non_blocking=True,
+        experimental_thread_pool=futures.ThreadPoolExecutor(
+            max_workers=_THREAD_POOL_SIZE))
+
+    health_pb2_grpc.add_HealthServicer_to_server(health_servicer, server)
+    SERVICE_NAMES = (
+        test_pb2.DESCRIPTOR.services_by_name["TestService"].full_name,
+        health_pb2.DESCRIPTOR.services_by_name["Health"].full_name,
+        channelz_pb2.DESCRIPTOR.services_by_name["Channelz"].full_name,
+        reflection.SERVICE_NAME,
+    )
+    for service in SERVICE_NAMES:
+        health_servicer.set(service, health_pb2.HealthCheckResponse.SERVING)
+    reflection.enable_server_reflection(SERVICE_NAMES, server)
+
+
+def _configure_test_server(server: grpc.Server, port: int, secure_mode: bool,
+                           server_id: str) -> None:
+    test_pb2_grpc.add_TestServiceServicer_to_server(
+        TestService(server_id, socket.gethostname()), server)
+    listen_address = f"{_LISTEN_HOST}:{port}"
+    if not secure_mode:
```

Comment:
Just curious, will we test other kinds of fallback credentials?

Reply:
This is not currently part of the security interop test spec. However, we do have more coverage on that in unit tests.